### PR TITLE
[Serve] Improve the replica recover scenario

### LINKED ISSUE #35851: Release tests long_running_serve_failure.aws failed
### What happened + What you expected to happen

Release tests long_running_serve_failure.aws failed on release 2.5.0 branch, and I think on master too
- https://buildkite.com/ray-project/release-tests-branch/builds/1705#_

### Versions / Dependencies

master
release 2.5.0

### Reproduction script

buildkite

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- Remove the dead lock scenario in the recover step.
- Add default value for Deployment property.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sihanwang41** [comment]
The release tests passed (in my workspace). Rerunning to make sure it is consistent.

@edoakes can you take a look?

**Comment by edoakes** [comment]
@sihanwang41 please add a unit test to `test_deployment_state.py`

File: python/ray/serve/_private/replica.py [hunk 1]
```diff
@@ -196,14 +196,17 @@ async def initialize_replica():
                     controller_handle,
                     app_name,
                 )
-                self._init_finish_event.set()
+                self._init_finish = True
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:199]: I think this should be `self._replica_init_finish`

File: python/ray/serve/_private/replica.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ async def __init__(
             )
 
             # Indicates whether the replica has finished initializing.
-            self._init_finish_event = asyncio.Event()
+            self._replica_init_finish = False
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:161]:
```suggestion
            self._initialized = False
```
no need to prefix it with "replica" given this whole file is `replica.py` :)

File: python/ray/serve/_private/replica.py [hunk 3]
```diff
@@ -311,8 +314,12 @@ async def is_initialized(
             # Unused `_after` argument is for scheduling: passing an ObjectRef
             # allows delaying reconfiguration until after this call has returned.
             try:
-                await self._initialize_replica()
-                metadata = await self.reconfigure(deployment_config)
+                async with self._replica_init_lock:
+                    if not self._replica_init_finish:
+                        await self._initialize_replica()
+                if deployment_config:
+                    await self.reconfigure(deployment_config)
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:321]: do we need to worry about concurrency issues with running `reconfigure`?

File: python/ray/serve/_private/replica.py [hunk 4]
```diff
@@ -664,7 +669,10 @@ async def prepare_for_shutdown(self):
         while True:
             # Sleep first because we want to make sure all the routers receive
             # the notification to remove this replica first.
-            await asyncio.sleep(self.deployment_config.graceful_shutdown_wait_loop_s)
+            if self.deployment_config:
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:672]: why's this necessary?

File: release/long_running_tests/workloads/serve_failure.py [hunk 5]
```diff
@@ -69,13 +69,28 @@ def update_progress(result):
 
 @ray.remote(max_restarts=-1, max_task_retries=-1)
 class RandomKiller:
+
+    # The interval is to make sure controller will be able to broadcast routes
+    # and recover replicas.
+    CONTROLLER_KILL_INTERVAL = 20
```
Comment (edoakes) [inline release/long_running_tests/workloads/serve_failure.py:75]: this is a totally orthogonal change, if we include this it's not clear if the other code changes actually solve the problem (the test was running fine before the offending PR).

File: python/ray/serve/_private/replica.py (hunk 1 above)
Comment (zcin) [inline python/ray/serve/_private/replica.py:199]: Doesn't look like this is used anywhere - should this be `_replica_init_finish`?

File: python/ray/serve/_private/replica.py (hunk 3 above)
Comment (zcin) [reply python/ray/serve/_private/replica.py:321]: Since this call to reconfigure is something that is meant to be called once directly following replica initialization, would it be safer to just place it inside the lock?

File: python/ray/serve/_private/replica.py (hunk 4 above)
Comment (sihanwang41) [reply python/ray/serve/_private/replica.py:672]: Eyeballing the log and find this error(might not related to the release test), because self.deployment_config is initialized to be None in the constructor, if no `start` to be called, and then being called to stop, it might cause None type error.

File: python/ray/serve/_private/replica.py (hunk 3 above)
Comment (sihanwang41) [reply python/ray/serve/_private/replica.py]: good point

Comment (sihanwang41) [reply python/ray/serve/_private/replica.py]: Moved under the lock cloure.

File: python/ray/serve/_private/replica.py (hunk 4 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:672]: got it. we should still use some short default here else this will basically busy-spin

**Comment by edoakes** [comment]
Synced with @sihanwang41 offline. Plan is to run one more batch with the bugfix but without the interval change & try to diagnose the issue.

File: python/ray/serve/_private/replica.py [hunk 6]
```diff
@@ -311,8 +314,12 @@ async def is_initialized(
             # Unused `_after` argument is for scheduling: passing an ObjectRef
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/_private/replica.py:314]: :) done

File: python/ray/serve/_private/replica.py [hunk 7]
```diff
@@ -196,14 +198,17 @@ async def initialize_replica():
...
+                self._initialized = True
 
             # Is it fine that replica is None here?
             # Should we add a check in all methods that use self.replica
             # or, alternatively, create an async get_replica() method?
             self.replica = None
             self._initialize_replica = initialize_replica
 
+            # Used to make sure only one single init task happen.
+            self._replica_init_lock = asyncio.Lock()
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:210]:
```suggestion
            # Used to guard `initialize_replica` so that it isn't called twice.
            self._replica_init_lock = asyncio.Lock()
```

File: python/ray/serve/_private/replica.py [hunk 8]
```diff
@@ -303,16 +308,20 @@ async def is_allocated(self) -> str:
                 get_component_logger_file_path(),
             )
 
-        async def is_initialized(
+        async def initialized_and_get_metadata(
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:311]:
```suggestion
        async def initialize_and_get_metadata(
```

File: python/ray/serve/_private/replica.py [hunk 9]
```diff
@@ -303,16 +308,20 @@ async def is_allocated(self) -> str:
...
             deployment_config: DeploymentConfig = None,
             _after: Optional[Any] = None,
-        ):
+        ) -> Tuple[DeploymentConfig, DeploymentVersion]:
             # Unused `_after` argument is for scheduling: passing an ObjectRef
             # allows delaying reconfiguration until after this call has returned.
             try:
-                await self._initialize_replica()
-                metadata = await self.reconfigure(deployment_config)
+                async with self._replica_init_lock:
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:319]: Add a comment here for why we need the locking behavior (in what scenario this could be called twice).

File: python/ray/serve/_private/replica.py [hunk 10]
```diff
@@ -303,16 +308,20 @@ async def is_allocated(self) -> str:
...
+                    if not self._initialized:
+                        await self._initialize_replica()
+                    if deployment_config:
+                        await self.reconfigure(deployment_config)
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:323]: it's a behavior change to guard `reconfigure` with `if deployment_config`, please remove that (unless it's part of the fix?)

File: python/ray/serve/_private/replica.py (hunk 8 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:311]: "initialized" is the past participle, use `initialize` here because this function is what does the initialization

**Review by edoakes — COMMENTED** [review]
Only one minor comment about `reconfigure` behavior change + some nits, then LGTM.

**Comment by edoakes** [comment]
@sihanwang41 I think Java test failure may be relevant as well

File: python/ray/serve/_private/replica.py (hunk 10 above)
Comment (sihanwang41) [reply python/ray/serve/_private/replica.py:323]: In the recover step, we are not passing `deployment_config`.  Previously we are calling `_get_metadata`, unifying replica `recover` & `start` calling the same function, that is why we have the if check here.

File: python/ray/serve/_private/replica.py [hunk 11]
```diff
@@ -196,14 +198,17 @@ async def initialize_replica():
...
-                self._init_finish_event.set()
+                self._initialized = True
 
             # Is it fine that replica is None here?
             # Should we add a check in all methods that use self.replica
             # or, alternatively, create an async get_replica() method?
             self.replica = None
             self._initialize_replica = initialize_replica
 
+            # Used to guard `initialize_replica` so that it isn't called twice.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica.py:209]: [Nit] Can you explain why `initialize_replica` would get called twice?

File: python/ray/serve/tests/test_controller_recovery.py [hunk 12]
```diff
@@ -249,5 +252,53 @@ def get_actor_info(name: str):
...
+            replica_state = list(ReplicaState)
+        else:
+            replica_state = [replica_state]
+        return replicas.get(replica_state)
+
+    # Make sure the replica is in STOPPING state.
+    wait_for_condition(lambda: len(check_replica(ReplicaState.STOPPING)) > 0)
+    ray.kill(serve.context._global_client._controller, no_restart=False)
+    # Make sure the replica is in STOPPING state.
+    wait_for_condition(lambda: len(check_replica(ReplicaState.STOPPING)) > 0)
```
Comment (zcin) [inline python/ray/serve/tests/test_controller_recovery.py:292]: Should we check that the replica state is stopping for a few extra iterations to be extra safe? To remove the possibility that the replica state is only briefly set to stopping after the controller recovers, and make sure it's actually stuck there until we unblock the signal.

**Review by zcin — COMMENTED** [review]
The fix LGTM, once tests pass

File: python/ray/serve/_private/replica.py (hunk 11 above)
Comment (sihanwang41) [reply python/ray/serve/_private/replica.py:209]: When the controller dead, controller will try to initialize the replica again. Basically with this change, initialize_and_get_metadata will become idempotent, controller can call this function during start and recover without worrying about concurrency issue.

File: python/ray/serve/_private/replica.py (hunk 10 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:325]: Got it, thanks for the explanation

**Review by edoakes — APPROVED** [review]
LGTM please fix DCO so I can merge
